src/index.js

```diff
@@ -84,7 +86,10 @@ export const useElementScroll = ({
   }
 }
 
-const useWindowRect = (windowRef, initialRect = { width: 0, height: 0 }) => {
+export const useWindowRect = (
```

Comment:
can you add the below to lines 107-108?

```
        width: element.innerWidth ? element.innerWidth : element.offsetWidth,
        height: element.innerHeight ? element.innerHeight : element.offsetHeight,
```

When I add this and your changes in this PR, it allows me to use any outer container, not just Window, because innerWidth/height only exist on window and not other elements

The change makes it some 'outerSize' calculates correctly, so it knows what the start/end indexes are

Reply:
also, thank you for this!  This update came right when I needed it :)

Reply:
I actually ended up having to add one more thing to the `scrollToFn` in `useWindowScroll`:

```
    const scrollElementKey = horizontal ? 'scrollLeft' : 'scrollTop'
    // grab different properties if an element and not 'window'
    const refScrollKey = windowRef.current[scrollKey] === undefined ? scrollElementKey : scrollKey
    const refRectKey = windowRef.current[rectKey] === undefined ? scrollElementKey : rectKey

    const delta = ['ToIndex', 'SizeChanged'].includes(reason) ? windowRef.current[refScrollKey] + parentRef.current.getBoundingClientRect()[rectKey] : 0
    windowRef.current.scrollTo({ [refRectKey]: offset + delta })
```

Reply:
and if anyone else stumbles into this somehow - I had to swap all of the `useLayoutEffect` that were doing `setElement` based on ref.current with just `useEffect`

This was due to using React Suspense - the ref.currents were null on the useLayoutEffect, but somehow weren't on the useEffect - I do not know why, but I think React 18 will solve that - didn't seem like there were any issues or performance differences using useEffect on these either

Reply:
@bryan-hunter thanks, imho we should merge this first, then add new changes. 

This looks interesting 

>  I had to swap all of the useLayoutEffect that were doing setElement based on ref.current with just useEffect

Could you create a codesnabdox with the bug and add new issue.

Reply:
@piecyk 
turns out the useEffect vs useLayoutEffect only helps with a workaround for the real issue (and still doesn't solve the main issue most of the time)

here's a sandbox that reproduces it:

https://codesandbox.io/s/toend-forked-cphpc?file=/src/index.js

Basically, when a component suspends, the fallback component is renderered and that height is 0 due to display 'none' right when it executes.

If the fallback component were exactly the same height as what the eventually rendered component was going to be, I think it'd work, but that's hard to determine especially since this is specifically for dynamic mode and estimateSize is just an estimate and not exact.

Here's the main react issue that talks about it: https://github.com/facebook/react/issues/14536 - so it seems like concurrent mode will solve this, but in the meantime is very tough to troubleshoot and find a workable solution for

Reply:
@bryan-hunter hmm in your example even if post are fetched the component is not rendered that's why list is empty as there are not posts 🤔 

Forked example from Dan https://reactjs.org/docs/concurrent-mode-suspense.html#start-fetching-early, looks ok after adding useVirtual https://codesandbox.io/s/with-suspense-wyt0w 🤷‍♂️

Reply:
@piecyk the example I was going for had a Suspense at the Row level in the virtual list

the example that you're using has suspense wrapping the virtual list

it's only an issue when the row suspends, which is probably the most common use case when using dynamic mode

Reply:
@piecyk this is interesting, I modified your sandbox to next the ProfileDetails per row and made it's fetch time be 2000 ms so it was longer than the posts

https://codesandbox.io/s/with-suspense-forked-3bqcs?file=/src/index.js

when I set the fallback div height to "100%" i get maximum call depth exceeded
if I set the fallback div height to "10px" it renders based on that, and then it updates to real height after scrolling a bit

That sandbox was also using react next (18)

Reply:
@bryan-hunter yeah, update the same stuff 👍 https://codesandbox.io/s/with-suspense-wyt0w but used with 5000, hmm why you need the `fallback div height to "100%"` ? Looking on my version it's not needed. 

> it updates to real height after scrolling a bit

This is expected as hook don't know that row size was changed, it's not tracked automatically, one workaround is to trigger parent to re-render (where the hook is called ) or use something to observe it, like ResizeObserver checkout this #28

Reply:
first off, thanks for responding and being so helpful on everything in this repo :)

I think I am seeing where my thinking was flawed.

I use a fallback skeleton that is 100% width/height for most things like this - and that was assuming the wrapper container was giving it the space it needed

On my Fixed lists, it worked great because the estimateSize was always a real, fixed size so the container wrappers always provided the skeletons the right space to fill to 100%.

the Dynamic lists are actually the opposite - the container can estimate the size and provide it, but ultimately we're asking the library to re-measure it once it renders because the child knows its own height, and that's where the Suspense issue was coming in.

I'm actually wondering...if I just let each row's Suspense carry all the way up above where the measureRef is, maybe it won't execute the measureRef until it's done suspending and has a height.. 🤔 🤔

Reply:
eureka - https://codesandbox.io/s/toend-forked-cphpc?file=/src/index.js

that seemed to work, just had to make sure <Suspense> and the promise were executed above the <div> with the measureRef

Reply:
ended up going with the ResizeObserver - too many other things were suspending on the same page above the list on initial render and no guarantees on display none not messing with the height of a row

I believe I will also need this ResizeObserver on my "container ref" that isn't the windowRef - but again I feel like that is only necessary because of Suspense and the display none.

Thanks for all of the help and making the library so flexible to handle these anomalies - it's even flexible enough to handle React issues :p

Reply:
If you want to check out this Library used heavily in action with a lot of infinite scrolls, check out the RYSE app :p - https://apps.apple.com/us/app/ryse/id1560237752 - it's new and free and wouldn't be nearly as smooth without this nice library

Reply:
@bryan-hunter glad to help, awesome that it worked for you!